File: ports/lua/test/policy_test.lua
local h = require("test_helper")
local policy = require("unicode_lua.security.policy")

local ROOT = "testdata"

local function scan_case(case)
  return policy.scan(case.profile, case.mode, case.input)
end

local function scan_encoded_case(case)
  local scanners = {
    ["utf-8"] = policy.scan_utf8,
    ["utf-16be"] = policy.scan_utf16be,
    ["utf-16le"] = policy.scan_utf16le,
    ["utf-32be"] = policy.scan_utf32be,
    ["utf-32le"] = policy.scan_utf32le,
  }
  return scanners[case.encoding](case.profile, case.mode, case.input_bytes)
end

local function assert_required(case, verdict)
  local codes = h.codes(verdict)
  for _, required in ipairs(case.required_findings) do
    h.assert_includes(codes, required, case.name)
  end
  local by_code = {}
  for _, finding in ipairs(verdict.findings) do
    by_code[finding.code] = finding.positions
  end
  for _, expected in ipairs(case.required_positions or {}) do
    h.assert_equal(by_code[expected.code], expected.positions, case.name)
  end
end

local payload = h.json_file(ROOT .. "/fixtures/security/policy_contract.json")
for _, case in ipairs(payload.cases) do
  local verdict = scan_case(case)
  h.assert_equal(verdict.action, case.action, case.name)
  assert_required(case, verdict)
end

payload = h.json_file(ROOT .. "/fixtures/security/verdict_contract.json")
for _, case in ipairs(payload.cases) do
  local verdict = scan_case(case)
  h.assert_equal(policy.verdict_to_wire(verdict), case.verdict, case.name)
end

payload = h.json_file(ROOT .. "/fixtures/security/decode_contract.json")
for _, case in ipairs(payload.cases) do
  local verdict = policy.scan_utf8(case.profile, case.mode, case.input_bytes)
  h.assert_equal(verdict.action, case.action, case.name)
  h.assert_equal(verdict.input, case.input, case.name)
  assert_required(case, verdict)
end

payload = h.json_file(ROOT .. "/fixtures/security/decode_multiencoding_contract.json")
for _, case in ipairs(payload.cases) do
  local verdict = scan_encoded_case(case)
  h.assert_equal(verdict.action, case.action, case.name)
  h.assert_equal(verdict.input, case.input, case.name)
  assert_required(case, verdict)
end

File: ports/elixir/test/policy_test.exs
defmodule UnicodeSecurity.PolicyTest do
  use ExUnit.Case, async: false
  alias UnicodeSecurity.Policy
  import UnicodeSecurity.TestHelpers

  test "reason codes are stable" do
    assert Policy.reason_code(:tag_block_payload, "DirectAscii") ==
             "unicode.security.C.tag-block-payload.DirectAscii"

    assert Policy.reason_code(:bidi_control_balance) ==
             "unicode.security.C.bidi-control-balance.hazard"

    assert Policy.reason_code(:homoglyph_confusable, "TargetMatch") ==
             "unicode.security.I.homoglyph-confusable.TargetMatch"

    assert Policy.reason_code(:mixed_script_admissibility, "CrossScriptMix") ==
             "unicode.security.I.mixed-script-admissibility.CrossScriptMix"

    assert Policy.reason_code(:noncharacter_control, "Noncharacter") ==
             "unicode.security.C.noncharacter-control.Noncharacter"

    assert Policy.reason_code(:malformed_utf8, "InvalidStartByte") ==
             "unicode.security.C.malformed-utf8.InvalidStartByte"
  end

  test "policy contract fixture" do
    payload = fixture_json("policy_contract.json")
    assert payload["contract"] == "unicode-security-policy-v0"

    Enum.each(payload["cases"], fn case_data ->
      verdict = Policy.scan(case_data["profile"], case_data["mode"], case_data["input"])
      assert verdict.action == case_data["action"], case_data["name"]
      codes = codes(verdict)

      Enum.each(case_data["required_findings"], fn required ->
        assert required in codes, case_data["name"]
      end)
    end)
  end

  test "verdict contract fixture" do
    payload = fixture_json("verdict_contract.json")
    assert payload["contract"] == "unicode-security-verdict-v0"

    Enum.each(payload["cases"], fn case_data ->
      verdict = Policy.scan(case_data["profile"], case_data["mode"], case_data["input"])
      assert Policy.verdict_to_wire(verdict) == case_data["verdict"], case_data["name"]

      assert Policy.verdict_to_json(verdict) == JSON.encode!(case_data["verdict"]),
             case_data["name"]
    end)
  end

  # Agreement with the reference over a generated input stream. The corpus shares
  # the verdict contract's schema, so it runs through the same comparison, but
  # its cases come from the Rust reference over a deterministic stream rather
  # than being hand-written: agreement here is evidence that this port decides as
  # the reference does on inputs nobody chose, across every profile.
  test "differential corpus" do
    payload = fixture_json("differential_corpus.json")
    assert payload["contract"] == "unicode-security-verdict-v0"

    Enum.each(payload["cases"], fn case_data ->
      verdict = Policy.scan(case_data["profile"], case_data["mode"], case_data["input"])
      assert Policy.verdict_to_wire(verdict) == case_data["verdict"], case_data["name"]

      assert Policy.verdict_to_json(verdict) == JSON.encode!(case_data["verdict"]),
             case_data["name"]
    end)
  end

  test "utf-8 decode contract fixture" do
    payload = fixture_json("decode_contract.json")
    assert payload["contract"] == "unicode-security-decode-v0"

    Enum.each(payload["cases"], fn case_data ->
      verdict =
        Policy.scan_utf8(case_data["profile"], case_data["mode"], case_data["input_bytes"])

      assert verdict.action == case_data["action"], case_data["name"]
      assert verdict.input == case_data["input"], case_data["name"]
      assert_required_findings_and_positions(case_data, verdict)
    end)
  end

  test "multi-encoding decode contract fixture" do
    payload = fixture_json("decode_multiencoding_contract.json")
    assert payload["contract"] == "unicode-security-multiencoding-decode-v0"

    Enum.each(payload["cases"], fn case_data ->
      verdict = scan_encoded_case(case_data)
      assert verdict.action == case_data["action"], case_data["name"]
      assert verdict.input == case_data["input"], case_data["name"]
      assert_required_findings_and_positions(case_data, verdict)
    end)
  end
end
